### PR TITLE
memcheck for C SDK

### PR DESCRIPTION
**Pull request by ericwol-msft**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `main` branch. 
  - [ ] I have merged the latest `main` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

**Review by vaavva — APPROVED** [review]

File: iothub_client/src/iothub_client_core_ll.c [hunk 1]
```diff
@@ -975,24 +978,27 @@ static IOTHUB_CLIENT_CORE_LL_HANDLE_DATA* initialize_iothub_client(const IOTHUB_
...
                             free(result);
                             result = NULL;
                         }
-                        else if ((IoTHubSuffix = (char*)malloc(suffix_len + 1)) == NULL)
+                        else if (malloc_size_hub_suffix == SIZE_MAX ||
+                            (IoTHubSuffix = (char*)malloc(malloc_size_hub_suffix)) == NULL)
                         {
                             LogError("unable to malloc");
                             IoTHubClient_Auth_Destroy(result->authorization_module);
+                            free(IoTHubName);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply iothub_client/src/iothub_client_core_ll.c:996]: this was a bug fix. if the 991 malloc fails, the hub calloc would be leaked

File: iothub_client/src/iothubtransport_amqp_connection.c [hunk 2]
```diff
@@ -165,15 +166,15 @@ static int create_connection_handle(AMQP_CONNECTION_INSTANCE* instance)
         connection_io_transport = instance->underlying_io_transport;
     }
 
-    if ((unique_container_id = (char*)malloc(sizeof(char) * DEFAULT_UNIQUE_ID_LENGTH + 1)) == NULL)
+    size_t calloc_size = safe_multiply_size_t(sizeof(char), safe_add_size_t(DEFAULT_UNIQUE_ID_LENGTH, 1));
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply iothub_client/src/iothubtransport_amqp_connection.c:169]: this was a bug fix. the + 1 was for the NULL char. UUID len plus null, and then multiply the size of char.

File: iothub_client/src/iothubtransport_amqp_messenger.c [hunk 3]
```diff
@@ -337,15 +338,15 @@ static STRING_HANDLE create_link_name(role link_role, const char* device_id)
     char* unique_id;
     STRING_HANDLE result;
 
-    if ((unique_id = (char*)malloc(sizeof(char) * UNIQUE_ID_BUFFER_SIZE + 1)) == NULL)
+    size_t calloc_size = safe_multiply_size_t(sizeof(char), safe_add_size_t(UNIQUE_ID_BUFFER_SIZE, 1));
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply iothub_client/src/iothubtransport_amqp_messenger.c:341]: yup bug fix

File: provisioning_client/adapters/hsm_client_riot.c [hunk 4]
```diff
@@ -599,15 +600,16 @@ char* hsm_client_riot_get_certificate(HSM_CLIENT_HANDLE handle)
...
         size_t total_len = x509_client->alias_cert_length + x509_client->device_signed_length;
 
-        result = (char*)malloc(total_len + 1);
-        if (result == NULL)
+        size_t malloc_size = safe_add_size_t(total_len, 1);
+        if (malloc_size == SIZE_MAX ||
+            (result = (char*)malloc(malloc_size)) == NULL)
         {
-            LogError("Failed to allocate cert buffer.");
+            LogError("Failed to allocate cert buffer, size:%zu", malloc_size);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply provisioning_client/adapters/hsm_client_riot.c:607]: interesting, the compliers did not complain about this.

File: iothub_client/src/iothubtransport_amqp_telemetry_messenger.c [hunk 5]
```diff
@@ -269,14 +270,14 @@ static STRING_HANDLE create_link_name(const char* prefix, const char* infix)
     char* unique_id;
     STRING_HANDLE tag = NULL;
 
-    if ((unique_id = (char*)malloc(sizeof(char) * UNIQUE_ID_BUFFER_SIZE + 1)) == NULL)
+    size_t calloc_size = safe_multiply_size_t(sizeof(char), safe_add_size_t(UNIQUE_ID_BUFFER_SIZE, 1));
+    if (calloc_size == SIZE_MAX ||
+        (unique_id = (char*)calloc(1, calloc_size)) == NULL)
     {
-        LogError("Failed generating an unique tag (malloc failed)");
+        LogError("Failed generating an unique tag (malloc failed), size:%zu", calloc_size);
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_amqp_telemetry_messenger.c:277]:
```suggestion
        LogError("Failed generating an unique tag (calloc failed), size:%zu", calloc_size);
```

File: iothub_client/src/iothubtransport_amqp_twin_messenger.c [hunk 6]
```diff
@@ -341,14 +342,15 @@ static char* generate_unique_id(void)
 {
     char* result;
 
-    if ((result = (char*)malloc(sizeof(char) * UNIQUE_ID_BUFFER_SIZE + 1)) == NULL)
+    size_t calloc_size = safe_multiply_size_t(sizeof(char), safe_add_size_t(UNIQUE_ID_BUFFER_SIZE, 1));
+    if (calloc_size == SIZE_MAX ||
+        (result = (char*)calloc(1, calloc_size)) == NULL)
     {
-        LogError("Failed generating an unique tag (malloc failed)");
+        LogError("Failed generating an unique tag (malloc failed), size:%zu", calloc_size);
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_amqp_twin_messenger.c:349]:
```suggestion
        LogError("Failed generating an unique tag (calloc failed), size:%zu", calloc_size);
```

File: serializer/inc/serializer_devicetwin.h [hunk 7]
```diff
@@ -52,10 +53,12 @@ static void serializer_ingest(DEVICE_TWIN_UPDATE_STATE update_state, const unsig
...
     int result;
-    char* payloadZeroTerminated = (char*)malloc(size + 1);
-    if (payloadZeroTerminated == NULL)
+    char* payloadZeroTerminated;
+    size_t malloc_size = safe_add_size_t(size, 1);
+    if (malloc_size == SIZE_MAX ||
+        (payloadZeroTerminated = (char*)malloc(malloc_size)) == NULL)
     {
-        LogError("failure in malloc");
+        LogError("failure in malloc, size:%zu", malloc_size);
```
Comment (ewertons) [inline serializer/inc/serializer_devicetwin.h:61]:
```suggestion
        LogError("deviceMethodCallback failure allocating payloadZeroTerminated, size:%zu", malloc_size);
```

**Review by ewertons — APPROVED** [review]

**Pull request merged**